dali/operators/reader/nvdecoder/dynlink_nvcuvid.cc

```diff
@@ -1,163 +1,88 @@
+  if (!ret) {
+    printf("Failed to find required function \"%s\" in %s\n", name.c_str(), __DriverLibName);
   }
+  return ret;
+}
 
-#define CHECKED_CALL(call)          \
-  do {                              \
-    CUresult result = (call);       \
-    if (CUDA_SUCCESS != result) {   \
-      return result;                \
-    }                               \
-  } while(0)
-
-#define GET_PROC_REQUIRED(name) GET_PROC_EX(name,name,1)
-#define GET_PROC_OPTIONAL(name) GET_PROC_EX(name,name,0)
-#define GET_PROC(name)          GET_PROC_REQUIRED(name)
-#define GET_PROC_V2(name)       GET_PROC_EX_V2(name,name,1)
-
-CUresult cuvidInit(unsigned int Flags, DLLDRIVER &driver_lib)
-{
-  driver_lib = NULL;
-  CHECKED_CALL(LOAD_LIBRARY(&driver_lib));
-
-  // fetch all function pointers
-  GET_PROC(cuvidCreateVideoSource);
-  GET_PROC(cuvidCreateVideoSourceW);
-  GET_PROC(cuvidDestroyVideoSource);
-  GET_PROC(cuvidSetVideoSourceState);
-  GET_PROC(cuvidGetVideoSourceState);
-  GET_PROC(cuvidGetSourceVideoFormat);
-  GET_PROC(cuvidGetSourceAudioFormat);
-
-  GET_PROC(cuvidCreateVideoParser);
-  GET_PROC(cuvidParseVideoData);
-  GET_PROC(cuvidDestroyVideoParser);
-
-
-  GET_PROC(cuvidGetDecoderCaps);
-  GET_PROC(cuvidCreateDecoder);
-  GET_PROC(cuvidDestroyDecoder);
-  GET_PROC(cuvidDecodePicture);
-  GET_PROC_OPTIONAL(cuvidGetDecodeStatus);
-  GET_PROC_OPTIONAL(cuvidReconfigureDecoder);
-
-#if !defined(__CUVID_DEVPTR64) || defined(__CUVID_INTERNAL)
-  GET_PROC(cuvidMapVideoFrame);
-  GET_PROC(cuvidUnmapVideoFrame);
-#endif
-
-#if defined(_WIN64) || defined(__LP64__) || defined(__x86_64) || defined(AMD64) || defined(_M_AMD64)
-  GET_PROC(cuvidMapVideoFrame64);
-  GET_PROC(cuvidUnmapVideoFrame64);
-#endif
-  GET_PROC(cuvidCtxLockCreate);
-  GET_PROC(cuvidCtxLockDestroy);
-  GET_PROC(cuvidCtxLock);
-  GET_PROC(cuvidCtxUnlock);
-
-  return CUDA_SUCCESS;
 }
 
+// it is defined in the generated file
+typedef void *tLoadSymbol(const std::string &name);
+void NvcuvidSetSymbolLoader(tLoadSymbol loader_func);
+void Nvcuvid_2SetSymbolLoader(tLoadSymbol loader_func);
+
 DLLDRIVER cuvidInitChecked(unsigned int Flags) {
-  DLLDRIVER driver_lib;
-  std::lock_guard<std::mutex> lock(m);
-  cuvidInit(Flags, driver_lib);
-  return driver_lib;
+  std::lock_guard<std::mutex> lock(nvcuvid_mutex);
+  if (!nvcuvidDrvLib) {
+    nvcuvidDrvLib = loadNvcuvidLibrary();
+  }
+
+  // set symbol loader for this library
+  NvcuvidSetSymbolLoader(LoadSymbol);
+  Nvcuvid_2SetSymbolLoader(LoadSymbol);
+  return nvcuvidDrvLib;
 }
 
 void cuvidDeinit(DLLDRIVER driver_lib) {
-  std::lock_guard<std::mutex> lock(m);
+  std::lock_guard<std::mutex> lock(nvcuvid_mutex);
   dlclose(driver_lib);
+  nvcuvidDrvLib = nullptr;
+}
+
+bool cuvidIsSymbolAvailable(const std::string &name) {
+  static std::mutex symbol_mutex;
+  static std::unordered_map<std::string, void*> symbol_map;
+  std::lock_guard<std::mutex> lock(symbol_mutex);
+  auto it = symbol_map.find(name);
+  if (it == symbol_map.end()) {
```

Comment:
I don't like that a function called "is symbol availave" is modying the symbol map. Can you explain why is that?

Reply:
It lazy loads the symbols to the map only for the purpose of checking is it is available. It is not very frequently used functionality, so it won't happen very often.
I don't have a good idea how to move it to the generated file and not affect the perf.

Reply:
Current implementation of generated functions is:
```
  using FuncPtr = CUresult (CUDAAPI *)(CUresult, const char **);
  static auto func_ptr = reinterpret_cast<FuncPtr>(load_symbol_func("cuGetErrorString"));
  if (!func_ptr) return CUDA_ERROR_SHARED_OBJECT_SYMBOL_NOT_FOUND;
  return func_ptr(error, pStr);
}
```
If we add a map there we would need add a sync point for that.
We cannot easily expose the f_pointer itself as well from it to check the value.
No idea how to approach differently.

Reply:
I was just commenting on the name. This function is actually loading symbols. Perhaps it should be called something like `LoadSymbolIfNeeded` to better reflect what it does

Reply:
But it doesn't load the symbol either, in the way that the first invocation of the relevant function will call again `LoadSymbol`.
I would keep the name as it is as this is the purpose of the function - check if a function with the given name is available, not to load it. One day we may rework the stub implementation but the external API would rather stay the same.